### DISCUSSION TITLE
fixing STL files using opencascade

### DISCUSSION THREAD

**Discussion by hrgdavor** [donalffons/opencascade.js#77] [Q&A] [answered]
I have STLs that report non manifold vertices or get broken in prusa slicer. I have noticed that there is shape healing, https://dev.opencascade.org/doc/overview/html/occt_user_guides__shape_healing.html

... and I am wondering if I could use opencascade to try to fix some of them.

**Comment by donalffons** [comment, marked as answer]
Hey!
I personally haven't used any of OpenCascade's shape healing functions, but I would definitely be interested to hear about your experiences, in case you're giving it a shot.

I have used [`BRepBuilderAPI_Sewing`](https://dev.opencascade.org/doc/refman/html/class_b_rep_builder_a_p_i___sewing.html) in OpenCascade.js before, for loading tessellated 3D geometries and attempting to convert them into solids. That approach did seem to work ok, at least with my limited testing. My approach with `BRepBuilderAPI_Sewing` was this:

1. Construct a `TopoDS_Compound` and throw all the individual triangles (constructed as `BRepBuilderAPI_MakeFace` from `BRepBuilderAPI_MakePolygon` from `BRepBuilderAPI_MakeVertex`) into it. 
2. Create a `BRepBuilderAPI_Sewing`, set your desired tolerances and `::Load()` the compound
3. Run `::Perform()` on the sewing instance.
4. Get your sewed shape via `::SewedShape()`

The shape healing approach seems to offer much more fine-grained control than my simplistic attempt.

**Reply by hrgdavor** [reply]
found this git repo just today, and I am curious about the healing functions. I might have few more noob questions to get the basic script working. 

Thanks for the hints, much appreciated, also thanks for the quick reply.

**Reply by donalffons** [reply]
Feel free to shoot your questions! I'll try to help.

**Comment by sridhar-mani** [comment]
can anyone tell whats the proper way to load stl or step files into vtk canvas or three js canvas using opencascade
